regression/run/references/main.f

```diff
@@ -58,10 +59,24 @@ subroutine test_arraywrapper
     call assert_true(associated(arr4, arr))
     call assert_equals(10, size(arr4))
 
-    call arrinst%fetch_array(arr5)
+    call arrinst%fetch_array_ptr(arr5)
     call assert_true(associated(arr5, arr))
     call assert_equals(10, size(arr5))
-    
+
+    call arrinst%fetch_array_ref(arr6)
+    call assert_true(associated(arr6, arr))
+    call assert_equals(10, size(arr6))
+
+    voidptr = C_NULL_PTR
+    call arrinst%fetch_void_ptr(voidptr)
+    call assert_true(c_associated(voidptr, c_loc(arr)), "fetchVoidPtr")
```

Comment:
What can you do with `voidptr` in Fortran (i.e. the result of `arrinst%fetch_void_ptr(voidptr)`)?

Specifically, I'm assuming Fortran doesn't know the size of what `voidptr` is pointing to (since this is not set). 
Can you:
* do indirection on it in Fortran?
* access the `n`th item?
* pass it to a C function that expects a pointer?

Reply:
Mostly it's useful to have a producer and consumer of some data structure which is too complex for Fortran to deal with, ex 'double ***'.
Like in C, the only way to dereference it is to 'cast' it via `c_f_pointer`. With that function, from iso_c_bindings module, you can set the type and shape.

Reply:
Thanks @ltaylor16 -- might be useful to add that use case to the example `main.f` and/or docs?

Reply:
The next line does that:

    call assert_true(arrinst%check_ptr(voidptr), "checkPtr")

`check_ptr` compares the `tyep(C_PTR)`  against the address in the instance.